torch/csrc/jit/codegen/cuda/lower_insert_syncs.cpp

```diff
@@ -0,0 +1,192 @@
+            // Child - WAR race
+            // Parent first sync
+            // Inherit Child Initial / Clear Parent Final
+            initial_sync_ = true;
+            last_op_sync_ = true;
+            initial_.insert(child.initial_.begin(), child.initial_.end());
+            final_.clear();
+          } else {
+            // Child - 1+
+            // Parent first sync
+            // Inherit Child Initial + Final
+            initial_sync_ = true;
+            initial_.insert(child.initial_.begin(), child.initial_.end());
+            final_.insert(child.final_.begin(), child.final_.end());
+          }
+        } else {
+          // Parent - 1+
+          if (!child.initial_sync_) {
+            // Child - None
+            // Append All Child to Parent Last
+            final_.insert(child_inputs.begin(), child_inputs.end());
+          } else if (child.war_sync_) {
+            // Child - WAR race
+            // Clear Parent Last / Discard Child Initial
+            last_op_sync_ = true;
+            final_.clear();
+          } else {
+            // Child - 1+
+            // Inherit Child Final / Discard Child Initial
+            final_.insert(child.final_.begin(), child.final_.end());
+          }
+        }
+      } else {
+        handle(expr);
+      }
+    }
+
+    if (!fl->iter_domain()->isThread() && !fl->iter_domain()->isBroadcast()) {
+      // Determine if any smem TV is written to at beginning of the for-loop
+      // and whether that smem TV is read from at the end of the for-loop
+      std::vector<TensorView*> common_data;
+      std::set_intersection(
+          initial_.begin(),
+          initial_.end(),
+          final_.begin(),
+          final_.end(),
+          std::back_inserter(common_data),
+          [](TensorView* left, TensorView* right) {
+            return left->name() == right->name();
+          });
+
+      // Insert new SyncThreads at end of for-loop to prevent WAR race condition
+      if (common_data.size() > 0 &&
+          fl->body().exprs().back()->getExprType().value() != ExprType::Sync &&
+          !last_op_sync_) {
+        // std::cout << "WAR race detected; Add Sync" << std::endl;
+        war_sync_ = true;
+        fl->body().push_back(new kir::Sync());
+      }
+    }
+  }
+
+  void hasOutputSmemExpr(Expr* expr, std::unordered_set<TensorView*>& set) {
+    for (auto out : expr->outputs()) {
+      if (ir_utils::isTV(out)) {
+        auto tv = out->as<TensorView>();
+        if (tv->getMemoryType() == MemoryType::Shared) {
+          set.insert(tv);
+        }
+      }
+    }
+  }
+
+  void hasInputSmemExpr(Expr* expr, std::unordered_set<TensorView*>& set) {
+    for (auto inp : expr->inputs()) {
+      if (ir_utils::isTV(inp)) {
+        auto tv = inp->as<TensorView>();
+        if (tv->getMemoryType() == MemoryType::Shared) {
+          set.insert(tv);
+        }
+      }
+    }
+  }
+
+ private:
+  explicit LocalSyncInserter(Expr* expr) {
+    handle(expr);
```

Comment:
Is this called for all expressions recursively? Does redundant analysis happen here? Looks like the traversal cost would be exponential to the number of expressions. Can this be done in O(N)?

Reply:
I am performing recursive depth-first search on the nested for-loop structure. Since it is a DAG, each for-loop is only visited once.

Each LocalSyncInserter only handles a single for-loop. The expressions in the nested for-loop are not revisited by the parent LocalSyncInserter.

Reply:
Hmm, but this is called for all expressions at `SyncInserter::modify`. Each expression is visited once in a single call from `SyncInserter::modify`, but any expressions nested in other expressions such as arithmetic expressions in a loop seem to be visited multiple times, doesn't it?

Reply:
`LocalSyncInserter` is called for all expressions in `const std::vector<Expr*>& exprs`, not every expression in the kernel. `exprs` only contains nested for-loops based on the computeAt structure. The expressions in the for-loops are not repeated in `const std::vector<Expr*>& exprs`.

Reply:
Ah, thanks! That's what I misunderstood.